presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java

```diff
@@ -831,9 +900,103 @@ public long getTimeSpentWaitingForOutputInMillis()
+    {
+        private static final MutablePartitionId DEFAULT_PARTITION = new MutablePartitionId();
+
+        private final PrestoSparkOutputBuffer<PrestoSparkBufferedSerializedPage> outputBuffer;
+        private final TempStorage tempStorage;
+        private final TempDataOperationContext tempDataOperationContext;
+        private final long maxBufferedBytes;
+        private TempDataSink tempDataSink;
+
+        private long timeSpentWaitingForOutputInMillis;
+
+        private DiskPageOutputSupplier(PrestoSparkOutputBuffer<PrestoSparkBufferedSerializedPage> outputBuffer,
+                TempStorage tempStorage,
+                TempDataOperationContext tempDataOperationContext,
+                long maxBufferedBytes)
+        {
+            this.outputBuffer = requireNonNull(outputBuffer, "outputBuffer is null");
+            this.tempStorage = requireNonNull(tempStorage, "tempStorage is null");
+            this.tempDataOperationContext = requireNonNull(tempDataOperationContext, "tempDataOperationContext is null");
+            this.maxBufferedBytes = requireNonNull(maxBufferedBytes, "pageFileStripeMaxSize is null");
+        }
+
+        @Override
+        public Tuple2<MutablePartitionId, PrestoSparkDiskPage> getNext()
+                throws InterruptedException
+        {
+            long start = System.currentTimeMillis();
+            PrestoSparkBufferedSerializedPage page = outputBuffer.get();
+            if (page == null) {
+                return null;
+            }
+
+            int compressedBroadcastSizeInBytes = 0;
+            int uncompressedBroadcastSizeInBytes = 0;
+            int positionCount = 0;
+            List<Long> pageChecksum = new ArrayList<>();
+            TempStorageHandle tempStorageHandle;
+            try {
+                this.tempDataSink = tempStorage.create(tempDataOperationContext);
+                List<DataOutput> bufferedPages = new ArrayList<>();
+                long bufferedBytes = 0;
+
+                while (page != null) {
+                    PageDataOutput pageDataOutput = new PageDataOutput(page.getSerializedPage());
+                    long writtenSize = pageDataOutput.size();
+
+                    if ((maxBufferedBytes - bufferedBytes) < writtenSize) {
+                        tempDataSink.write(bufferedPages);
+                        bufferedPages.clear();
+                        bufferedBytes = 0;
+                    }
+
+                    bufferedPages.add(pageDataOutput);
+                    bufferedBytes += writtenSize;
+                    compressedBroadcastSizeInBytes += page.getSerializedPage().getSizeInBytes();
+                    uncompressedBroadcastSizeInBytes += page.getSerializedPage().getUncompressedSizeInBytes();
+                    positionCount += page.getPositionCount();
+                    pageChecksum.add(computeChecksum(page.getSerializedPage().getSlice().byteArray()));
+                    page = outputBuffer.get();
```

Comment:
outputBuffer will return null if the task finished unsuccessfully. Ideally we should remove temporary file if the task has failed. Task failure is checked in the `PrestoSparkTaskExecutor#doComputeNext` ( see `if (failures.isEmpty()) {`)